src/Orleans.Testing.Utils/GrainFactoryMocker.cs

```diff
@@ -275,5 +315,39 @@ public GrainFactoryMocker GuidCompoundGrain<TGrain>(Guid? primaryKey = null, str
 
             return this;
         }
+
+        public GrainFactoryMocker LongCompoundGrain<TGrain>(long? primaryKey = null, string compoundKey = null, Action<Mock<TGrain>> mocker = null, Action<long, string> callback = null)
+            where TGrain : class, IGrainWithIntegerCompoundKey
+        {
+            var mockGrain = new Mock<TGrain>();
+            mockGrain.DefaultValue = DefaultValue.Mock;
+
+            if (!_mockGrainWithLongCompoundKeyRegistry.TryGetValue(typeof(TGrain), out var registry))
+            {
+                registry = new Dictionary<(long?, string), Mock>();
+                _mockGrainWithLongCompoundKeyRegistry.Add(typeof(TGrain), registry);
+            }
+
+            var key = (primaryKey, compoundKey);
+
+            // We've already mocked this grain
+            if (registry.ContainsKey(key)) throw new Exception("LongCompoundGrain already mocked - you must call ClearLongCompoundGrain() first");
+
+            registry.Add(key, mockGrain);
+            mocker?.Invoke(mockGrain);
+
+            Mock
+                .Setup(x => x.GetGrain<TGrain>(
+                    It.Is<long>(g => primaryKey.HasValue ? g == primaryKey : true),
+                    It.Is<string>(s => !string.IsNullOrEmpty(compoundKey) ? compoundKey.Equals(s, StringComparison.InvariantCulture) : true),
+                    It.IsAny<string>()))
+                .Callback((long g, string k, string c) => callback?.Invoke(g, k))
+                .Returns(() =>
+                {
+                    return mockGrain.Object;
+                });
+
+            return this;
+        }
```

Comment:
Might be worthwhile to pull in the storage provider action waitable stuff added in the sharding changes. /shrug